hugr-core/src/export.rs

```diff
@@ -344,20 +327,101 @@ impl<'a> Context<'a> {
                     regions = self.bump.alloc_slice_copy(&[region]);
                 }
 
-                model::Operation::Custom { operation }
+                model::Operation::CustomFull { operation }
             }
         };
 
-        let signature = self.module.insert_term(model::Term::Wildcard);
+        let (signature, num_inputs, num_outputs) = match optype {
+            OpType::DataflowBlock(block) => {
+                let signature = self.export_block_signature(block);
+                (Some(signature), 1, block.sum_rows.len())
+            }
 
-        self.module.insert_node(model::Node {
+            // PERFORMANCE: As it stands, `OpType::dataflow_signature` copies and/or allocates.
+            // That might not seem like a big deal, but it's a significant portion of the time spent
+            // when exporting. However it is not trivial to change this at the moment.
+            _ => match &optype.dataflow_signature() {
+                Some(signature) => {
+                    let num_inputs = signature.input_types().len();
+                    let num_outputs = signature.output_types().len();
+                    let signature = self.export_func_type(signature);
+                    (Some(signature), num_inputs, num_outputs)
+                }
+                None => (None, 0, 0),
+            },
+        };
+
+        let inputs = self.make_ports(node, Direction::Incoming, num_inputs);
+        let outputs = self.make_ports(node, Direction::Outgoing, num_outputs);
+
+        // Replace the placeholder node with the actual node.
+        *self.module.get_node_mut(node_id).unwrap() = model::Node {
             operation,
             inputs,
             outputs,
             params,
             regions,
             meta: &[], // TODO: Export metadata
             signature,
+        };
+
+        node_id
+    }
+
+    /// Export the signature of a `DataflowBlock`. Here we can't use `OpType::dataflow_signature`
+    /// like for the other nodes since the ports are control flow ports.
```

Comment:
does this do what you need: https://github.com/CQCL/hugr/blob/a725a29a13648ffa262b53e1ce8dd605711c0540/hugr-core/src/ops/controlflow.rs#L165

Reply:
Not for here: this constructs the outer signature of a dataflow block, not the inner. We construct the inner signature from the input and output nodes in the dfg region export.